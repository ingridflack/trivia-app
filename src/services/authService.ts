import axios from "axios";
import { BASE_URL } from "../config/baseUrl";

interface LoginParams {
  email: string;
  password: string;
}

export const login = ({ email, password }: LoginParams) =>
  axios.post(`${BASE_URL}/login`, { email, password });
