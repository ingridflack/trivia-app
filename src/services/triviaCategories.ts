import axios from "../config/api";
import { Category, ApiResponse } from "../types/sharedTypes";

interface CategoryResponse extends ApiResponse {
  categories: Category[];
}

export const getCategories = () =>
  axios.get<CategoryResponse>("/trivia/categories");
