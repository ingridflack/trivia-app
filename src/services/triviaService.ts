import axios from "../config/api";
import { Category, ApiResponse, Trivia } from "../types/sharedTypes";

interface CategoryResponse extends ApiResponse {
  categories: Category[];
}

interface TriviaQueryParams {
  category: number;
  difficulty: string;
  amount: number;
  type?: string;
}

interface TriviaResponse extends ApiResponse {
  triviaId: string;
}

interface TriviaByIdResponse extends ApiResponse {
  trivia: Trivia;
}

export const getCategories = () =>
  axios.get<CategoryResponse>("/trivia/categories");

export const createTrivia = (params: TriviaQueryParams) =>
  axios.post<TriviaResponse>(
    `/trivia?amount=${params.amount}&type=&difficulty=${params.difficulty}&category=${params.category}`
  );

export const getTriviaById = (triviaId: string) =>
  axios.get<TriviaByIdResponse>(`/trivia/${triviaId}`);
