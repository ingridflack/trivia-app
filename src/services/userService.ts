import axios from "../config/api";
import { UserSearch } from "../types/sharedTypes";

export const search = (params: Partial<UserSearch>) =>
  axios.get<UserSearch[]>("/users/search", { params });

export const getUser = () => axios.get("/users");
