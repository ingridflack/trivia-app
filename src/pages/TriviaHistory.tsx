import { useEffect, useState } from "react";
import * as TriviaService from "../services/triviaService";
import { TriviaHistory as TriviaHistoryInterface } from "../types/sharedTypes";
import {
  Accordion,
  AccordionButton,
  AccordionIcon,
  AccordionItem,
  AccordionPanel,
  Box,
  Container,
  List,
  ListItem,
  Tag,
  TagLabel,
  Text,
} from "@chakra-ui/react";
import Menu from "../components/Menu";
import DOMPurify from "dompurify";

const CATEGORY_LABELS: { [key: string]: string } = {
  9: "General Knowledge",
  10: "Entertainment: Books",
  31: "Entertainment: Japanese Anime & Manga",
  19: "Science: Mathematics",
  14: "Entertainment: Television",
};

export default function TriviaHistory() {
  const [triviaHistory, setTriviaHistory] = useState<TriviaHistoryInterface[]>(
    []
  );

  useEffect(() => {
    const fetchTriviaHistory = async () => {
      const { data } = await TriviaService.getTriviaHistory();

      console.log(data.triviaHistory);

      setTriviaHistory(data.triviaHistory);
    };

    fetchTriviaHistory();
  }, []);

  console.log(triviaHistory);

  return (
    <div>
      <Menu />

      <Container
        centerContent
        height="100%"
        alignItems="center"
        justifyContent="center"
      >
        <List w="100%">
          {triviaHistory.map((historyItem) => (
            <ListItem key={historyItem.trivia._id}>
              <Accordion allowMultiple>
                <AccordionItem>
                  <h2>
                    <AccordionButton>
                      <Box as="span" flex="1" textAlign="left">
                        {CATEGORY_LABELS[historyItem.trivia.category] ||
                          "Mixed"}{" "}
                        | {historyItem.trivia.difficulty} |{" "}
                        <strong>
                          <small>Points: </small>
                        </strong>
                        {historyItem.trivia.score || 0}
                      </Box>
                      <AccordionIcon />
                    </AccordionButton>
                  </h2>
                  <AccordionPanel pb={4} pr={16}>
                    {historyItem.items.map((item) => (
                      <Box
                        display="flex"
                        justifyContent="space-between"
                        flexDirection="column"
                        mb={4}
                        gap={4}
                      >
                        <Text
                          fontSize="medium"
                          dangerouslySetInnerHTML={{
                            __html: DOMPurify.sanitize(item.question.question),
                          }}
                        />
                        <Box display="flex" alignItems="center" gap="4">
                          <Text fontSize="small">
                            <strong>Answer time: </strong>
                            {item.answerTime} seconds
                          </Text>

                          <Tag
                            size={"md"}
                            key={item._id}
                            borderRadius="full"
                            variant="solid"
                            colorScheme={item.isCorrect ? "green" : "red"}
                            w="fit-content"
                          >
                            <TagLabel>
                              {" "}
                              <Text fontSize="small" fontWeight="bold">
                                {item.isCorrect ? "Correct" : "Incorrect"}
                              </Text>
                            </TagLabel>
                          </Tag>
                        </Box>
                      </Box>
                    ))}
                  </AccordionPanel>
                </AccordionItem>
              </Accordion>
            </ListItem>
          ))}
        </List>
      </Container>
    </div>
  );
}
