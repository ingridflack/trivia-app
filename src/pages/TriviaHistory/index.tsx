import { useEffect, useState } from "react";
import * as TriviaService from "../../services/triviaService";
import { TriviaHistory as ITriviaHistory } from "../../types/sharedTypes";
import {
  Accordion,
  AccordionButton,
  AccordionIcon,
  AccordionItem,
  AccordionPanel,
  Badge,
  Box,
  Container,
  Grid,
  Heading,
  Skeleton,
  Stack,
  Text,
  Tooltip,
} from "@chakra-ui/react";
import Header from "../../components/Header";
import DOMPurify from "dompurify";
import { badgeColor } from "../../helpers/trivia";
import { CheckIcon, CloseIcon } from "@chakra-ui/icons";
import { TRIVIA_CATEGORIES } from "../../constants/trivia";
import { Footer } from "../../components/Footer";
import { Link } from "react-router-dom";

interface TriviaHistoryContentProps {
  triviaHistory: ITriviaHistory[];
  isLoading: boolean;
}

function TriviaHistoryContent({
  triviaHistory,
  isLoading,
}: TriviaHistoryContentProps) {
  if (isLoading) {
    return (
      <Stack>
        <Skeleton height="40px" width="100%" />
        <Skeleton height="40px" width="100%" />
        <Skeleton height="40px" width="100%" />
        <Skeleton height="40px" width="100%" />
      </Stack>
    );
  }

  if (!triviaHistory.length) {
    return (
      <Text>
        Your trivia history is empty.{" "}
        <Text
          as={Link}
          to="/trivia/config"
          textDecoration="underline"
          fontWeight="700"
        >
          Click here
        </Text>{" "}
        to start a game!
      </Text>
    );
  }

  return (
    <Accordion allowMultiple>
      {triviaHistory
        .filter((item) => item.completed)
        .map((historyItem) => (
          <AccordionItem
            key={historyItem.trivia._id}
            backgroundColor="white"
            marginBottom="10px"
          >
            <AccordionButton>
              <Box
                as="span"
                flex="1"
                textAlign="left"
                color="gray.500"
                display="flex"
                justifyContent="space-between"
                flexDirection={{ base: "column", md: "row" }}
                gap={{ base: "10px", md: "0" }}
              >
                <Box display="flex" alignItems="center" gap="16px">
                  <Tooltip
                    label={historyItem.trivia.status}
                    textTransform="capitalize"
                  >
                    <Box
                      backgroundColor={
                        historyItem.trivia.status === "completed"
                          ? "green.500"
                          : "orange.500"
                      }
                      borderRadius="full"
                      width="10px"
                      height="10px"
                    />
                  </Tooltip>
                  <Text>
                    {TRIVIA_CATEGORIES[Number(historyItem.trivia.category)]
                      ?.title || "General Knowledge "}
                  </Text>
                </Box>

                <Box display="flex" alignItems="center" gap="30px">
                  <Text fontSize="sm">
                    Score: {historyItem.score || 0} / {historyItem.items.length}
                  </Text>
                  <Badge
                    display="flex"
                    alignItems="center"
                    justifyContent="center"
                    colorScheme={badgeColor(historyItem.trivia.difficulty)}
                  >
                    {historyItem.trivia.difficulty}
                  </Badge>
                  <Text fontSize="sm">
                    {new Date(
                      historyItem.trivia.createdAt
                    ).toLocaleDateString()}
                  </Text>
                  <AccordionIcon />
                </Box>
              </Box>
            </AccordionButton>
            <AccordionPanel padding={0}>
              <Grid
                templateColumns={{ base: "1fr", md: "repeat(2, 1fr)" }}
                borderTop="1px solid var(--chakra-colors-gray-100)"
              >
                {historyItem.items.map((item) => (
                  <Box
                    display="flex"
                    alignItems="flex-start"
                    justifyContent="space-between"
                    gap={4}
                    padding="20px"
                    borderBottom="1px solid var(--chakra-colors-gray-100)"
                    key={item._id}
                    _odd={{
                      borderRight: "1px solid var(--chakra-colors-gray-100)",
                    }}
                  >
                    <Text
                      color="gray.500"
                      display="flex"
                      fontSize="medium"
                      dangerouslySetInnerHTML={{
                        __html: DOMPurify.sanitize(item.question.question),
                      }}
                    />

                    <Text
                      fontSize="sm"
                      color={item.isCorrect ? "green.500" : "red.500"}
                    >
                      {item.isCorrect ? <CheckIcon /> : <CloseIcon />}
                    </Text>
                  </Box>
                ))}
              </Grid>
            </AccordionPanel>
          </AccordionItem>
        ))}
    </Accordion>
  );
}

export default function TriviaHistory() {
  const [triviaHistory, setTriviaHistory] = useState<ITriviaHistory[]>([]);
  const [isLoading, setIsLoading] = useState(true);

  useEffect(() => {
    const fetchTriviaHistory = async () => {
      try {
        const { data } = await TriviaService.getTriviaHistory();
        setTriviaHistory(data.triviaHistory);
      } catch (error) {
        console.log(error);
      } finally {
        setIsLoading(false);
      }
    };

    fetchTriviaHistory();
  }, []);

  return (
    <>
      <Header />

      <Container as="main" padding={{ base: "40px 20px", md: "80px 0 120px" }}>
        <Box as="header" marginBottom="40px">
          <Heading fontSize="3xl">Trivia History</Heading>
        </Box>

        <TriviaHistoryContent
          triviaHistory={triviaHistory}
          isLoading={isLoading}
        />
      </Container>

      <Footer />
    </>
  );
}
