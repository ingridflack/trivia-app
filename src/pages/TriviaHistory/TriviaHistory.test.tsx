import { afterEach, describe, expect, it, Mock, vi } from "vitest";
import TriviaHistory from ".";
import {
  cleanup,
  render,
  waitForElementToBeRemoved,
} from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import * as TriviaService from "../../services/triviaService";

vi.mock("../../services/triviaService");

describe("TriviaHistory", () => {
  afterEach(() => {
    cleanup();
  });

  it("should match the snapshot when trivia history is present", async () => {
    (TriviaService.getTriviaHistory as Mock).mockResolvedValue({
      data: {
        triviaHistory: [
          {
            trivia: {
              _id: "trivia1",
              category: "1",
              createdAt: "2023-10-01T00:00:00Z",
              difficulty: "easy",
              score: 5,
              status: "completed",
            },
            items: [
              {
                _id: "item1",
                answerTime: 30,
                isCorrect: true,
                question: { question: "Question 1?" },
              },
              {
                _id: "item2",
                answerTime: 20,
                isCorrect: false,
                question: { question: "Question 2?" },
              },
            ],
            completed: true,
          },
        ],
      },
    });

    const { container, queryByText } = render(
      <MemoryRouter>
        <TriviaHistory />
      </MemoryRouter>
    );

    await waitForElementToBeRemoved(() =>
      queryByText(/Your trivia history is empty/)
    );

    expect(container).toMatchSnapshot();
  });

  it("should match the snapshot when trivia history is empty", async () => {
    (TriviaService.getTriviaHistory as Mock).mockResolvedValue({
      data: {
        triviaHistory: [],
      },
    });

    const { container, queryByText } = render(
      <MemoryRouter>
        <TriviaHistory />
      </MemoryRouter>
    );

    expect(queryByText(/Your trivia history is empty/)).toBeTruthy();
    expect(container).toMatchSnapshot();
  });
});
