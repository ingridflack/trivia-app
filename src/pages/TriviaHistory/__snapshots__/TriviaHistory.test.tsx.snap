// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TriviaHistory > should match the snapshot when trivia history is empty 1`] = `
<div>
  <header
    class="css-cyjglw"
  >
    <div
      class="chakra-container css-old1by"
    >
      <div
        class="chakra-stack css-2jbh02"
      >
        <a
          href="/"
        >
          <p
            class="chakra-text css-kkaoep"
          >
            Trivia
          </p>
        </a>
        <div
          class="css-j0ozid"
        >
          <a
            href="/login"
          >
             Sign in 
          </a>
          <a
            href="/sign-up"
          >
             Sign Up 
          </a>
        </div>
      </div>
    </div>
  </header>
  <main
    class="chakra-container css-1clgay9"
  >
    <header
      class="css-10ib5jr"
    >
      <h2
        class="chakra-heading css-1tapsqs"
      >
        Trivia History
      </h2>
    </header>
    <p
      class="chakra-text css-0"
    >
      Your trivia history is empty.
       
      <a
        class="chakra-text css-1pnc3tu"
        href="/trivia/config"
      >
        Click here
      </a>
       
      to start a game!
    </p>
  </main>
  <footer
    class="css-1s2lyti"
  >
    <div
      class="chakra-container css-0"
    >
      <div
        class="chakra-stack css-11h4xmw"
      >
        <a
          href="/"
        >
          <p
            class="chakra-text css-rekzoj"
          >
            Trivia
          </p>
        </a>
        <div
          class="css-0"
        >
          <strong
            class="chakra-heading css-kf3i1p"
          >
            Site map
          </strong>
          <div
            class="chakra-stack css-1ko7mmx"
          >
            <a
              href="/"
            >
              Home
            </a>
            <a
              href="/login"
            >
              Sign in
            </a>
            <a
              href="/sign-up"
            >
              Sign up
            </a>
          </div>
        </div>
        <div
          class="css-0"
        >
          <strong
            class="chakra-heading css-kf3i1p"
          >
            Social
          </strong>
          <div
            class="chakra-stack css-1ko7mmx"
          >
            <a
              href="https://www.linkedin.com/in/ingrid-flack/"
            >
              LinkedIn
            </a>
            <a
              href="https://github.com/ingridflack"
            >
              GitHub
            </a>
          </div>
        </div>
      </div>
      <div
        class="chakra-stack css-1y3yl0q"
      >
        <p
          class="chakra-text css-0"
        >
          © 2024 Trivia App
        </p>
        <p
          class="chakra-text css-0"
        >
          Developed by Ingrid Flack
        </p>
      </div>
    </div>
  </footer>
</div>
`;

exports[`TriviaHistory > should match the snapshot when trivia history is present 1`] = `
<div>
  <header
    class="css-cyjglw"
  >
    <div
      class="chakra-container css-old1by"
    >
      <div
        class="chakra-stack css-2jbh02"
      >
        <a
          href="/"
        >
          <p
            class="chakra-text css-kkaoep"
          >
            Trivia
          </p>
        </a>
        <div
          class="css-j0ozid"
        >
          <a
            href="/login"
          >
             Sign in 
          </a>
          <a
            href="/sign-up"
          >
             Sign Up 
          </a>
        </div>
      </div>
    </div>
  </header>
  <main
    class="chakra-container css-1clgay9"
  >
    <header
      class="css-10ib5jr"
    >
      <h2
        class="chakra-heading css-1tapsqs"
      >
        Trivia History
      </h2>
    </header>
    <div
      class="chakra-accordion css-0"
    >
      <div
        class="chakra-accordion__item css-1ohcrly"
      >
        <button
          aria-controls="accordion-panel-:r0:"
          aria-expanded="false"
          class="chakra-accordion__button css-ubg1nw"
          data-index="0"
          id="accordion-button-:r0:"
          type="button"
        >
          <span
            class="css-1fw0p1f"
          >
            <div
              class="css-j0ozid"
            >
              <div
                class="css-1rtijl6"
              />
              <p
                class="chakra-text css-0"
              >
                General Knowledge 
              </p>
            </div>
            <div
              class="css-111q8q3"
            >
              <p
                class="chakra-text css-g5hezf"
              >
                Score: 
                5
                 /
                 
                2
              </p>
              <span
                class="chakra-badge css-9w0iez"
              >
                easy
              </span>
              <p
                class="chakra-text css-g5hezf"
              >
                30/09/2023
              </p>
              <svg
                aria-hidden="true"
                class="chakra-icon chakra-accordion__icon css-8o5ky2"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
                  fill="currentColor"
                />
              </svg>
            </div>
          </span>
        </button>
        <div
          class="chakra-collapse"
          style="overflow: hidden; display: none; opacity: 0; height: 0px;"
        >
          <div
            aria-labelledby="accordion-button-:r0:"
            class="chakra-accordion__panel css-old1by"
            id="accordion-panel-:r0:"
            role="region"
          >
            <div
              class="css-1aanveu"
            >
              <div
                class="css-13wf3bv"
              >
                <p
                  class="chakra-text css-1qd0q1g"
                >
                  Question 1?
                </p>
                <p
                  class="chakra-text css-1j6t4zg"
                >
                  <svg
                    class="chakra-icon css-onkibi"
                    focusable="false"
                    viewBox="0 0 14 14"
                  >
                    <g
                      fill="currentColor"
                    >
                      <polygon
                        points="5.5 11.9993304 14 3.49933039 12.5 2 5.5 8.99933039 1.5 4.9968652 0 6.49933039"
                      />
                    </g>
                  </svg>
                </p>
              </div>
              <div
                class="css-13wf3bv"
              >
                <p
                  class="chakra-text css-1qd0q1g"
                >
                  Question 2?
                </p>
                <p
                  class="chakra-text css-1jv2xzm"
                >
                  <svg
                    class="chakra-icon css-onkibi"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M.439,21.44a1.5,1.5,0,0,0,2.122,2.121L11.823,14.3a.25.25,0,0,1,.354,0l9.262,9.263a1.5,1.5,0,1,0,2.122-2.121L14.3,12.177a.25.25,0,0,1,0-.354l9.263-9.262A1.5,1.5,0,0,0,21.439.44L12.177,9.7a.25.25,0,0,1-.354,0L2.561.44A1.5,1.5,0,0,0,.439,2.561L9.7,11.823a.25.25,0,0,1,0,.354Z"
                      fill="currentColor"
                    />
                  </svg>
                </p>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </main>
  <footer
    class="css-1s2lyti"
  >
    <div
      class="chakra-container css-0"
    >
      <div
        class="chakra-stack css-11h4xmw"
      >
        <a
          href="/"
        >
          <p
            class="chakra-text css-rekzoj"
          >
            Trivia
          </p>
        </a>
        <div
          class="css-0"
        >
          <strong
            class="chakra-heading css-kf3i1p"
          >
            Site map
          </strong>
          <div
            class="chakra-stack css-1ko7mmx"
          >
            <a
              href="/"
            >
              Home
            </a>
            <a
              href="/login"
            >
              Sign in
            </a>
            <a
              href="/sign-up"
            >
              Sign up
            </a>
          </div>
        </div>
        <div
          class="css-0"
        >
          <strong
            class="chakra-heading css-kf3i1p"
          >
            Social
          </strong>
          <div
            class="chakra-stack css-1ko7mmx"
          >
            <a
              href="https://www.linkedin.com/in/ingrid-flack/"
            >
              LinkedIn
            </a>
            <a
              href="https://github.com/ingridflack"
            >
              GitHub
            </a>
          </div>
        </div>
      </div>
      <div
        class="chakra-stack css-1y3yl0q"
      >
        <p
          class="chakra-text css-0"
        >
          © 2024 Trivia App
        </p>
        <p
          class="chakra-text css-0"
        >
          Developed by Ingrid Flack
        </p>
      </div>
    </div>
  </footer>
</div>
`;
